feat: implement aliased commands with `miden` + Toolchain 0.16.0

lima-limon-inc opened the pull request:
This PR includes implements the `miden` command functionality. This includes:
- `miden <COMPONENT>` Calling a miden component by name, and `miden` taking care of which executable to call. The actual resolution from component name -> executable name is achieved in conjunction with the Manifest. That means that the available components are not hardcoded in `midenup` itself, but rather resolved at run time, using the manifest to aid in executable name resolution.
-  `miden <ALIAS>`, which map to a hardcoded set of known commands. 
- `miden help` which displays a help message (with two additional variants also exiting).

The list of known aliases are:
- `miden account`
- `miden faucet`
- `miden new`
- `miden build`
- `miden test`
- `miden deploy`
- `miden call`
- `miden send`
- `miden simulate`

The manifest got the following updates:
- Now the components specify which type of file they install, be it a library or an executable (if not specified, it as assumed that it installs an executable whose name is equal to the component's name).

bitwalker commented on src/main.rs:
```diff
@@ -164,24 +168,77 @@ fn main() -> anyhow::Result<()> {
+                        .collect();
+                    let help_message = format!(
+                        "The Miden toolchain porcelain
+
+{} {} <COMPONENT>
+
+Available components:
+{}
+
+Help:
+  help                   Print this help message
+  <COMPONENT> help       Print <COMPONENTS>'s help message
+",
```
No worries! I think your approach here is logical, I just happened to have already been wondering how the `help` mechanism was going to work, so I've been mulling it over for a bit 🙂

lima-limon-inc commented:
> Firstly, I created a `MidenAliases` enum which holds all the known miden aliases. This was done to "bundle" the mapping from command name to executable name; seen [here](https://github.com/0xMiden/midenup/pull/43/files#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcR136-R233). Similarly, I also created a `MidenComponents` enum with the same objective in mind, seen [here](https://github.com/0xMiden/midenup/pull/43/files#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcR46-R100).

With this setup in place, where we handle each component and alias explicitly, couldn't we remove the `external subcommand` tag from the `Miden` variant ([here](https://github.com/0xMiden/midenup/blob/e274e27bf60ec98c9ae7317cd61fb7fbb251e0d7/src/main.rs#L40-L41)). 

Wouldn't that aid in parsing and help message generation?

bitwalker commented:
I think it may help to take a step back and look at a couple of goals I think we should aim for with the `miden` CLI:

* It should work across toolchain versions, so it can't make assumptions about what is available, or how specific commands are structured, with few exceptions (and those exceptions are temporary as we get everything aligned).
* We want to avoid hardcoding too much information into `miden` itself, instead relying primarily on conventions we define (UX guidelines for executables it delegates to, how help is expected to be requested, binary names, relationship between binaries and the components that contain them, etc.).
* When conventions are insufficient (or as is currently the case, we haven't gotten to the point where we're aligned across the project on those conventions yet), only then do we add special cases to `miden` itself. Even here, it would be preferable to keep such hardcoded details to an absolute minimum. Examples of this are the actual `miden` subcommands that we want to have at the top level - a number of them don't correspond to the conventions we've defined for resolving external subcommands, and so we need to map those top-level subcommands to a specific component (from which we can use the manifest to tell us what the actual binary name is) and optionally, the subcommand of that component's CLI (in cases where the top-level subcommand has a different name in the underlying executable). What this means is that in `miden`, we basically have three pieces of "hardcoded" information:
  1. The name of a top-level subcommand which needs a non-conventional mapping.
  2. The name of the component to which that subcommand will be delegated. We use this to determine two things using the current toolchain manifest: whether the component is currently installed (so we can warn the user if it is missing and inform them how to install it), and when installed, the name of the underlying executable to invoke (this is either derived using the `miden-<component_name>` convention, or stored in the manifest using the `installed_file` key (IIRC).
  3. Optionally, arguments to provide to the underlying executable. This again should be rarely, if ever, used once all of our CLIs are aligned on conventions - but for now we're more reliant on hardcoded custom mappings.
* The current toolchain manifest (and the channel manifest that describes everything that is available) should provide us with all of the necessary information to discover available components and commands (i.e. not all components necessarily correspond to a `miden` subcommand, but many do). I think currently we're probably missing some information in the channel manifest/current toolchain manifest to do this entirely, namely which components provide executables, and what those executables are.

One thing we should probably do to avoid prolonging the current situation, is to make changes to the various Miden project CLIs to adhere to the following conventions:

* Always support `--help` to request help. Optionally support `miden-<command> help <topic>` in cases where specific help topics might be requested. In general though, we'd want to always be able to request subcommand help by appending `--help`.
* All CLIs with subcommands must be built as multi-call binaries, so that we can create symlinks in the toolchain bin directory for each subcommand we wish to surface at the top-level `miden` CLI. This is how `midenup` itself is built, so that the same executable can be used both to provide both `midenup` and `miden` via the corresponding symlink. Such binaries can still be used "normally", i.e. we could technically invoke `miden` via `midenup` using `midenup miden [ARGS...]`, but there isn't really a reason to do so. Our channel manifest can then know what the installed executable for a component is (via `installed_file` or convention), and we would additionally record what symlinks to the `installed_file` should be created to provide all subcommands we wish to surface via `miden` directly.
* All CLIs should generally use `clap` with consistent usage of common options (e.g. `-v/--verbose` for requesting verbose output, `-h/--help` for requesting help, `--color/--no-color` for enabling/disabling color output explicitly, etc.) and arguments (e.g. when invoking a command on a file path, it should be the sole positional argument, things like that).

In essence, I'd like to make `miden` know as little about individual commands and components as possible, unless that information is obtained from one of the manifests. This ensures that as you switch between toolchain versions, `miden` automatically adapts to how things were structured in a given toolchain version.

lima-limon-inc commented:
Apologies for the late reply. I have already started working on a new iteration with your changes in mind; and I've already pushed some of these changes. As soon as it's ready for ready, I'll update the PR's status.

lima-limon-inc commented:
I've just pushed the final commit with the `miden` command rework. I've moved the functionality into a separate file called `miden_wrapper` (linked [here](https://github.com/0xMiden/midenup/pull/43/files#diff-1a813ad5fe30a92311d801c75e869aa961e8e4f3605c553bee04c9dc4bc6c4b2)) to improve readability in `main.rs` and also to isolate the changes from all the other merged PRs.

I've updated the PR's description with the new approach, but TLDR:
- The only hardcoded bits left are the mappings from alias -> (component_name, prefix_arguments). 
- The installed file type (with the respective name) is now saved in the manifest, and *not* hardcoded in.
    - This is then used to obtain the component's executable name.
- `miden help`'s message does *not* install the current toolchain if missing; but rather provides a generic help message à la `clap`
   - There are other variants of `help` which are toolchain sensitive and can thus trigger an install. 

Regarding the comments about the rework of the other miden CLI's, I'll create a tracking issue to keep tabs on it. I'll be opening PR's with the corresponding patches in the various repositories ASAP.

Do let me know if there's any functionality that's missing/wrong here. Thanks in advance!

lima-limon-inc commented:
Btw, I am currently working on a different  [PR](https://github.com/0xMiden/midenup/pull/73) that is tackling this functionality:
> All CLIs with subcommands must be built as multi-call binaries, so that we can create symlinks in the toolchain bin directory for each subcommand we wish to surface at the top-level miden CLI. This is how midenup itself is built, so that the same executable can be used both to provide both midenup and miden via the corresponding symlink. Such binaries can still be used "normally", i.e. we could technically invoke miden via midenup using midenup miden [ARGS...], but there isn't really a reason to do so. Our channel manifest can then know what the installed executable for a component is (via installed_file or convention), and we would additionally record what symlinks to the installed_file should be created to provide all subcommands we wish to surface via miden directly.

That is, moving the aliases from the executable to the manifest. I decided against adding that functionality here since this PR seemed big as is.

bitwalker reviewed (approved):
Great job!
